open-machine-learning-jupyter-book/deep-learning/difussion-model.ipynb

```diff
@@ -1 +1,1010 @@
t_index):\n","\n","    betas_t = extract(betas, t, x.shape)### betas_t = 1-alphas_t\n","    sqrt_one_minus_alphas_cumprod_t = extract(sqrt_one_minus_alphas_cumprod, t, x.shape)### square root of 1-alpha_t_bar\n","    sqrt_recip_alphas_t = extract(sqrt_recip_alphas, t, x.shape)### 1/square root of alpha_t\n","\n","    model_mean = sqrt_recip_alphas_t * (x - betas_t * model([x, t]) / sqrt_one_minus_alphas_cumprod_t)### equation 4 of algorithm 2 above\n","\n","    if t_index == 0:\n","        return model_mean\n","    else:\n","        posterior_variance_t = extract(posterior_variance, t, x.shape)### sigma_t\n","        noise = tf.random.normal(x.shape)\n","        return model_mean + tf.math.sqrt(posterior_variance_t) * noise\n","\n","imgs = []\n","img = tf.random.normal((64,IM_SHAPE[0],IM_SHAPE[1],IM_SHAPE[2]))\n","for i in reversed(range(0, TIME_STEPS)):### we go backwards from t = 1000 to t = 0\n","    print(i)\n","    img = p_sample(model,img,tf.fill((1,),i,), i)\n","    imgs.append(img)\n","\n","plt.figure(figsize = (16,16))\n","\n","for i in range(32):\n","    ax = plt.subplot(4,8, i+1)\n","    plt.imshow(np.array(imgs[999])[i])\n","    plt.axis(\"off\")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"executionInfo":{"elapsed":178410,"status":"ok","timestamp":1705751030229,"user":{"displayName":"qiongying fu","userId":"01979569471380266441"},"user_tz":-480},"id":"bZ7YU-LNf2Sn","outputId":"3637cd6d-26e0-4d3d-8536-d420ace5b550"},"outputs":[],"source":["random_index = 0\n","\n","fig = plt.figure()\n","ims = []\n","for i in range(TIME_STEPS):\n","    im = plt.imshow(np.array(imgs[i])[random_index], animated=True)\n","    ims.append([im])\n","    ims.append([im])\n","\n","animate = animation.ArtistAnimation(fig, ims, interval=5, blit=True, repeat_delay=1000)\n","animate.save('diffusion.gif')\n","plt.show()"]},{"cell_type":"markdown","metadata":{"id":"UJabZ9MEf2Sn"},"source":["## Your turn! 🚀\n","\n","Assignment - [Denoising difussion model](../assignments/deep-learning/difussion-model/denoising-difussion-model.ipynb)\n","\n","## Acknowledgments\n","\n","Thanks to [Kwangnam Yu](https://github.com/phykn) for creating the open-source project [diffusion_models_tutorial](https://github.com/phykn/diffusion_models_tutorial) and [kaggle](https://www.kaggle.com/) for creating the open-source course [Denoising Diffusion Models with TensorFlow](https://www.kaggle.com/code/folefac/denoising-diffusion-models-with-tensorflow#3.-The-Celeb-A-Dataset-%F0%9F%92%BE). They inspire the majority of the content in this chapter."]}],"metadata":{"accelerator":"GPU","colab":{"gpuType":"T4","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.18"}},"nbformat":4,"nbformat_minor":0}
```

Comment:
<p>Take the output push</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ocademy-ai/machine-learning/pull/751/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>